### PR TITLE
Support for CFI unwind info

### PR DESCRIPTION
**Pull request by kyulee1**
For Unix targeting CoreRT, this will provide platform specific unwind
info which is a dwarf format.
Unlike window’s UNWIND_INFO, we won’t encode the format within RyuJit
which is not only complex/error-prone but also will be slightly
different depending on platforms.
Instead, CFI pseudo instructions are encoded, which will be passed to
CoreRT/ObjectWriter whi will translate them to directly emit CFI
instructions in LLVM MC to establish frames.
Then LLVM MC will handle the details of laying out the  eh_frame
sections.
One a JitFlag is used to dynamically dispatch either Windows’s unwind
blob or this CFI instruction table. No JIT/EE interface change is needed
since we’ve already passing an opaque blob.
Initially when I looked at what Clang does, the prologue and the
sequence of CFI emissions are a bit different than Windows.
Since we will emit the same sequence of code with the same runtime that
we define, I assume this is unnecessary – I’ve verified the CFI sequence
here can work correctly with libunwind.
Basically we need only 3 operations – push reg is a combination of 1 and
2 below.
 1. Allocation – increase stack frame. Normally subtract esp in x64.
 2. Store – Copy a (callee save) register to stack (memory)
 3. SaveFP – Set frame pointer register
Since Windows operation is based on the relative value (all offsets,
etc are computed from the current point), I also use the similar form of
CFI instructions .
So, mostly one window’s code corresponds to two CFIs – we might optimize
this by aggregating allocation, but I’d like to keep the current
syntax/semantic same as Windows.
This results in a very simple transformation in par with Windows.

### CODE REVIEW

File: src/inc/corjit.h [hunk 1]
```diff
@@ -124,16 +124,13 @@ enum CorJitFlag
...
-#endif // MDIL
-
-#ifdef MDIL
     // Safe to overlap with CORJIT_FLG_MAKEFINALCODE below. Not used by the JIT, used internally by NGen only.
     CORJIT_FLG_MINIMAL_MDIL        = 0x00008000, // Generate MDIL code suitable for use to bind other assemblies.
 
     // Safe to overlap with CORJIT_FLG_READYTORUN below. Not used by the JIT, used internally by NGen only.
     CORJIT_FLG_NO_MDIL             = 0x00010000, // Generate an MDIL section but no code or CTL. Not used by the JIT, used internally by NGen only.
+#else
+    CORJIT_FLG_CFI_UNWIND          = 0x00004000, // Emit CFI unwind info
```
Comment (kyulee1) [inline src/inc/corjit.h]: Thanks to Pat who will expand more Jit flags, I use this unused flag, instead.

**Comment by kyulee1** [comment]
@dotnet/jit-contrib PTAL. The corresponding encoder for CFI is in LLILC repo -- https://github.com/dotnet/llilc/pull/1011

File: src/jit/unwindamd64.cpp [hunk 2]
```diff
@@ -19,6 +19,44 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
...
+    case REG_RCX: dwarfReg = 2; break;
+    case REG_RDX: dwarfReg = 1; break;
+    case REG_RBX: dwarfReg = 3; break;
+    case REG_RSP: dwarfReg = 7; break;
+    case REG_RBP: dwarfReg = 6; break;
+    case REG_RSI: dwarfReg = 4; break;
+    case REG_RDI: dwarfReg = 5; break;
+    default:
+        if (reg >= REG_XMM0) {
+            dwarfReg++; // shift by 1 due to dwarf_reg_ip
```
Comment (AndyAyersMS) [inline src/jit/unwindamd64.cpp]: Are these XMM encodings correct?

File: src/inc/clrnt.h [hunk 3]
```diff
@@ -853,6 +853,35 @@ typedef union _UNWIND_CODE {
     USHORT FrameOffset;
 } UNWIND_CODE, *PUNWIND_CODE;
 
+// CFI_CODE is 8 byte struct regardless of operation for simplicity.
```
Comment (BruceForstall) [inline src/inc/clrnt.h]: You should write this comment to describe more about what a "CFI code" is, and how it is used. Maybe reference a spec. Etc.

File: src/inc/clrnt.h [hunk 4]
```diff
@@ -853,6 +853,35 @@ typedef union _UNWIND_CODE {
...
+// ObjectWriter will parse it and invoke a proper MCStream API in LLVM.
+typedef struct _CFI_CODE {
+    UCHAR CodeOffset;       // Offset from the start of code the frame covers.
+    UCHAR OpCode;           // Opcode encoded as Operation
```
Comment (BruceForstall) [inline src/inc/clrnt.h]: This comment makes not sense to me. what does "as Operation" mean?

File: src/inc/clrnt.h [hunk 5]
```diff
@@ -853,6 +853,35 @@ typedef union _UNWIND_CODE {
...
+    SHORT DwarfReg;         // Dwarf register number. -1 is invalid.
```
Comment (BruceForstall) [inline src/inc/clrnt.h]: What are all the possible values here? Where are they defined?

File: src/inc/clrnt.h [hunk 6]
```diff
@@ -853,6 +853,35 @@ typedef union _UNWIND_CODE {
...
+    INT Offset;             // Offset
```
Comment (BruceForstall) [inline src/inc/clrnt.h]: This comment is useless.

File: src/inc/clrnt.h [hunk 7]
```diff
@@ -853,6 +853,35 @@ typedef union _UNWIND_CODE {
...
+} CFI_CODE, *PCFI_CODE;
+
+// This is from LLVM MCDwarf.h
```
Comment (BruceForstall) [inline src/inc/clrnt.h]: This structure should be fully commented. Don't just reference some non-local file.

File: src/inc/corjit.h (hunk 1 above)
Comment (BruceForstall) [inline src/inc/corjit.h]: why doesn't this just replace the CORJIT_FLG_UNUSED7 definition above, such that the 0x00004000 values are adjacent?

File: src/jit/compiler.h [hunk 8]
```diff
@@ -1176,6 +1176,10 @@ struct FuncInfoDsc
     BYTE                unwindCodes[offsetof(UNWIND_INFO, UnwindCode) + (0xFF*sizeof(UNWIND_CODE))];
     unsigned            unwindCodeSlot;
 
+#define MAX_CFICODE_INDEX   0x200
+    BYTE                cfiCodes[(MAX_CFICODE_INDEX * sizeof(CFI_CODE))];
+    unsigned            cfiCodeIndex;
+
```
Comment (BruceForstall) [inline src/jit/compiler.h]: Shouldn't this be under some kind of Unix #ifdef?

You've got an unnecessary set of extra parens.

Why is `MAX_CFICODE_INDEX` defined as that number?

File: src/jit/compiler.h [hunk 9]
```diff
@@ -6899,7 +6903,8 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
     void                unwindReserveFuncHelper(FuncInfoDsc* func, bool isHotCode);
     void                unwindEmitFuncHelper(FuncInfoDsc* func, void* pHotCode, void* pColdCode, bool isHotCode);
     UNATIVE_OFFSET      unwindGetCurrentOffset(FuncInfoDsc* func);
-
+    int                 mapRegNumToDwarfReg(regNumber reg);
+    void                createCfiCode(FuncInfoDsc* func, UCHAR codeOffset, UCHAR opcode, USHORT dwarfReg, INT offset = 0);
```
Comment (BruceForstall) [inline src/jit/compiler.h]: Same as above: why isn't this under a Unix #ifdef?

File: src/jit/compiler.h (hunk 8 above)
Comment (kyulee1) [inline src/jit/compiler.h]: I'm reluctant to introduce #ifdef although strictly it is not used for Windows. In fact, we could encode SEH CFI similar to this logic instead of creating UNWIND_BLOB ourselves.
MAX_CFICODE_INDEX is also used to assert we don't overflow when we create an CFI.

Comment (kyulee1) [inline src/jit/compiler.h]: Do we have concern about TP/code size because of this? Or Is it convention to squeeze down the code as possible?

File: src/jit/unwindamd64.cpp [hunk 10]
```diff
@@ -19,6 +19,44 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 
 #if defined(_TARGET_AMD64_)
 
+int Compiler::mapRegNumToDwarfReg(regNumber reg)
+{
+    assert(reg >= 0 && reg <= REG_XMM15);
+    int dwarfReg = reg;
+
+    switch (reg) {
```
Comment (BruceForstall) [inline src/jit/unwindamd64.cpp]: CC: { on separate line

File: src/jit/unwindamd64.cpp [hunk 11]
```diff
@@ -19,6 +19,44 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
...
+    case REG_RDX: dwarfReg = 1; break;
+    case REG_RBX: dwarfReg = 3; break;
+    case REG_RSP: dwarfReg = 7; break;
+    case REG_RBP: dwarfReg = 6; break;
+    case REG_RSI: dwarfReg = 4; break;
+    case REG_RDI: dwarfReg = 5; break;
+    default:
+        if (reg >= REG_XMM0) {
+            dwarfReg++; // shift by 1 due to dwarf_reg_ip
+        }
```
Comment (BruceForstall) [inline src/jit/unwindamd64.cpp]: CC: there should be a 'break' for all case labels.

File: src/jit/compiler.h (hunk 8 above)
Comment (BruceForstall) [inline src/jit/compiler.h]: Why wouldn't you define cfiCodes as `CFI_CODE cfiCodes[MAX_CFICODE_INDEX]` ?

Comment (kyulee1) [inline src/jit/compiler.h]: Good point. My mind stikcs to the same logic in unwindCodes.

File: src/jit/unwindamd64.cpp [hunk 12]
```diff
@@ -121,6 +161,8 @@ void Compiler::unwindPush(regNumber reg)
     unsigned int cbProlog = unwindGetCurrentOffset(func);
     noway_assert((BYTE)cbProlog == cbProlog);
     code->CodeOffset = (BYTE)cbProlog;
+
+    createCfiCode(func, cbProlog, OpAdjustCfaOffset, -1, 8);
```
Comment (BruceForstall) [inline src/jit/unwindamd64.cpp]: It looks like -1 is special for dwarfReg. You should add a `#define` like DWARF_REG_ILLEGAL or similar

File: src/jit/unwindamd64.cpp [hunk 13]
```diff
@@ -410,6 +463,54 @@ void DumpUnwindInfo(bool isHotCode, UNATIVE_OFFSET startOffset, UNATIVE_OFFSET e
...
+            case OpDefCfaRegister:
+                assert(offset == 0);
+                printf("    CodeOffset: 0x%02X Op: DefCfaRegister DwarfReg:0x%X\n", codeOffset, dwarfReg);
+                break;
+            case OpAdjustCfaOffset:
+                assert(dwarfReg == -1);
+                printf("    CodeOffset: 0x%02X Op: AdjustCfaOffset Offset:0x%X\n", codeOffset, offset);
+                break;
+            default:
+                printf("    Unrecognized CFI_CODE: 0x%IX\n", *(UINT64*)pCode);
```
Comment (BruceForstall) [inline src/jit/unwindamd64.cpp]: All cases require "break"

File: src/jit/compiler.h (hunk 8 above)
Comment (BruceForstall) [inline src/jit/compiler.h]: Why is `MAX_CFICODE_INDEX` 0x200? Why not 0x201? 0x1000?

**Comment by BruceForstall** [comment]
I don't like how this looks. You're creating both CFI and Windows unwind info all the time, but only using one. Why don't you have a completely separate implementation of the unwind APIs which compiled for CFI generation? That is, one unwindAllocStack for Windows, one for CFI. Do you need to choose between these dynamically? If so, then have the AMD64 unwindAllocStack() dynamically choose between unwindAllocStackWindows() and unwindAllocStackUnix() (for example). Sounds like a good use of function pointers or virtual functions.

File: src/inc/clrnt.h [hunk 14]
```diff
@@ -853,6 +853,35 @@ typedef union _UNWIND_CODE {
...
+enum OpType {
+    OpSameValue,
+    OpRememberState,
```
Comment (jkotas) [inline src/inc/clrnt.h]: It would be better for this structure to live in separate file, maybe that is local to the JIT.

File: src/inc/clrnt.h (hunk 7 above)
Comment (jkotas) [inline src/inc/clrnt.h]: Also, if you have copied this from LLVM - you need to be careful about doing the right thing wrt. license.

**Comment by kyulee1** [comment]
@BruceForstall Both CFI and Windows unwind code is not completely orthogonal. Windows unwind code/emission sequence is used to emit epilogue, etc. So, I'd like to keep Window's unwind code process intact. Here CFI emission is not complete blob in the sense it's not usual until it is laid out by ObjectWriter/LLVM. So, think this CFI emission is a code-generation mode (under CoreRT) which tides to  Window's code sequence,
Splitting unwind infö based on platform, e.g, WIndows/Unix is also confusing since we always use Windows unwind info on CoreCLR/Unix.
In short,
1. default mode (regular Jit mode on CLR), we process/publish Window Unwind Info.
2. under this flag (e.g, CoreRT/Unix), we process Window Unwind info (for consistent code-gen) as well as CFI unwind info, but publish the latter only.

I am also not a big fan of virtual function stuff, and I don't see much value on this simple case.

File: src/inc/clrnt.h (hunk 14 above)
Comment (kyulee1) [inline src/inc/clrnt.h]: Makes sense. Probably instead of copying OpType directly from LLVM, I may introduce my own enum just for 3 cases for now. This way we can avoid license concern as well as don't need to worry about syncing this definition when LLVM source is changed.

**Comment by pgavlin** [comment]
/cc @cmckinsey @russellhadley 

> we process Window Unwind info (for consistent code-gen)

I'm not sure what you mean by this. It looks like consistent code generation should be ensured by the unwind info interface (i.e. the various `Compiler::unwind*` functions that you've changed). I also think that we should only be generating one style of unwind info or the other.

> Do you need to choose between these dynamically?
> ...
> I am also not a big fan of virtual function stuff, and I don't see much value on this simple case

I'm not sure that the former question has been answered yet. This really boils down to "do we want to be able to build RyuJIT s.t. a single build may generate code for multiple target platforms". AFAIK this is not possible today, and I'd be hesitant to make significant moves in that direction without strong justification (e.g. easier cross compilation). Just to be clear, I am not advocating that we further bifurcate the JIT/EE interface using #ifdefs (which we unfortunately have already done): I'm just suggesting that we think hard before doing the opposite to the JIT implementation and measure afterwards.

In terms of the overall direction: I thought that we were intending to introduce our own CFI codes that were compatible with what is expressible in DWARF and Windows unwind information s.t. we could have RyuJIT emit format-agnostic codes. Is this still the plan? If so, did it just so happen that DWARF CFI is expressive enough to cover this design?

**Comment by BruceForstall** [comment]
I talked to @kyulee1 in person; he can elaborate what he'll do (or, just make the changes and update the PR).

@pgavlin: I'm not in the loop for CoreRT, but I thought we had no plans to change how we generate Windows unwind codes. non-Windows codes could be pseudo-codes that work broadly on non-Windows platforms. Also, it turns out for non-Windows, CoreCLR (but not CoreRT), we do need to generate Windows unwind codes. And a desire for single RyuJIT binary that works for CoreCLR or CoreRT per-platform then requires dynamically generating one form or the other.

**Comment by pgavlin** [comment]
> I thought we had no plans to change how we generate Windows unwind codes

That's certainly correct when targeting CoreCLR et. al. I think this is still up in the air for CoreRT, however.

> Also, it turns out for non-Windows, CoreCLR (but not CoreRT), we do need to generate Windows unwind codes

Yes, that's correct.

> And a desire for single RyuJIT binary that works for CoreCLR or CoreRT per-platform

That's what I was missing! Thanks.

File: src/jit/unwindamd64.cpp (hunk 2 above)
Comment (kyulee1) [inline src/jit/unwindamd64.cpp]: I validated this CFI assembly directive. Dwarf XMM0 starts with 16 instead of REG_XMM0 (15). Note actually since all xmm are volatile in Unix ABI, we don't save them at all so this part is actually not used, but leave it for consistency.

**Comment by kyulee1** [comment]
@BruceForstall As discussed, I updated PR. 
- cfi.h is introduced to contain cfi data structure which will be consumed by object writer as well.
- cfiCodes are now dynamically allocated using a vector.
- ifdefed-out to split Windows and CFI code path. Note UNIX build implies both paths which is dynamically selected by a Jit flag.

There are some disruptiveness and code duplication due to splitting functions (for Windows/CFI), but I think they are reasonably small.
I think ideally we want a unwinder object that is attached to compiler object so that we can easily specialize unwinder logic per targets/platforms. But this refactoring seems beyond the scope of this PR, which we can come back later to clean this up.

File: src/inc/cfi.h [hunk 15]
```diff
@@ -0,0 +1,31 @@
+//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
```
Comment (BruceForstall) [inline src/inc/cfi.h]: Stephen Toub changed all the header file license agreements yesterday, so you should update this to match the new standard.

File: src/inc/cfi.h [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+    short DwarfReg;          // Dwarf register number. 0~32 for x64.
+    int Offset;
+    _CFI_CODE(unsigned char codeOffset, unsigned char cfiOpcode,
+        short dwarfReg, int offset)
+        : CodeOffset(codeOffset)
+        , CfiOpCode(cfiOpcode)
+        , DwarfReg(dwarfReg)
+        , Offset(offset)
+    {}
+} CFI_CODE, *PCFI_CODE;
```
Comment (BruceForstall) [inline src/inc/cfi.h]: since you've added a constructor, why don't you use a more C++ -like syntax (and less old-style C syntax), e.g.:

struct CFI_CODE { ... };
typedef CFI_CODE* PCFI_CODE;

File: src/jit/compiler.h [hunk 17]
```diff
@@ -1176,6 +1176,10 @@ struct FuncInfoDsc
     BYTE                unwindCodes[offsetof(UNWIND_INFO, UnwindCode) + (0xFF*sizeof(UNWIND_CODE))];
     unsigned            unwindCodeSlot;
 
+#ifdef UNIX_AMD64_ABI
+    jitstd::vector<CFI_CODE> *cfiCodes;
```
Comment (BruceForstall) [inline src/jit/compiler.h]: coding convention: * is placed adjacent to type. https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-jit-coding-conventions.md#10.1

**Comment by BruceForstall** [comment]
I like the change. Modulo the few comments, it looks good to me.

File: src/jit/unwindamd64.cpp (hunk 2 above)
Comment (pgavlin) [inline src/jit/unwindamd64.cpp]: Nit: why not just use a static table of mappings? If that's not possible, then I think I'd prefer that each case simply returned rather than mutating a local and leaving the return to the end. Something like:

```
assert(reg >= 0 && reg <= REG_XMM15);

switch (reg)
{
case REG_RAX: return 0;
case REG_RCX: return 2;
case REG_RDX: return 1;
case REG_RBX: return 3;
case REG_RSP: return 7;
case REG_RBP: return 6;
case REG_RSI: return 4;
case REG_RDI: return 5;
default:
	return reg < REG_XMM0 ? reg : reg + 1; // shift by 1 due to dwarf_reg_ip
}
```

File: src/jit/unwindamd64.cpp [hunk 18]
```diff
@@ -75,6 +120,37 @@ void Compiler::unwindBegProlog()
...
+    unwindGetFuncLocations(func, true, &func->startLoc, &func->endLoc);
+
+    if (fgFirstColdBlock != nullptr)
+    {
+        unwindGetFuncLocations(func, false, &func->coldStartLoc, &func->coldEndLoc);
+    }
+
+    jitstd::allocator<void> allocator(getAllocator());
+
+    func->cfiCodes = new (allocate_any<CFICodeVector>(allocator), jitstd::placement_t()) CFICodeVector(allocator);
```
Comment (pgavlin) [inline src/jit/unwindamd64.cpp]: Do we have any estimate on how many CFI codes we'll typically need for a function? If so, can we give this vector a reasonable initial capacity?

File: src/jit/unwindamd64.cpp [hunk 19]
```diff
@@ -75,6 +120,37 @@ void Compiler::unwindBegProlog()
...
+
+    // There is only one prolog for a function/funclet, and it comes first. So now is
+    // a good time to initialize all the unwind data structures.
+
+    unwindGetFuncLocations(func, true, &func->startLoc, &func->endLoc);
+
+    if (fgFirstColdBlock != nullptr)
+    {
+        unwindGetFuncLocations(func, false, &func->coldStartLoc, &func->coldEndLoc);
+    }
```
Comment (pgavlin) [inline src/jit/unwindamd64.cpp]: Looks like the code above could be shared between `unwindBegPrologCFI` and `unwindBegPrologWIndows`.

File: src/jit/unwindamd64.cpp [hunk 20]
```diff
@@ -142,6 +233,32 @@ void Compiler::unwindPush(regNumber reg)
...
 
+#ifdef UNIX_AMD64_ABI
+void Compiler::unwindPushCFI(regNumber reg)
+{
+    assert(compGeneratingProlog);
+
+    FuncInfoDsc* func = funCurrentFunc();
+
+    unsigned int cbProlog = unwindGetCurrentOffset(func);
+    noway_assert((BYTE)cbProlog == cbProlog);
```
Comment (pgavlin) [inline src/jit/unwindamd64.cpp]: Similarly, the code above could be pushed up into `unwindPush`.

File: src/jit/unwindamd64.cpp [hunk 21]
```diff
@@ -532,8 +785,12 @@ void Compiler::unwindEmitFuncHelper(FuncInfoDsc* func, void* pHotCode, void* pCo
 {
     UNATIVE_OFFSET startOffset;
     UNATIVE_OFFSET endOffset;
-    DWORD unwindCodeBytes;
-    BYTE* pUnwindBlock;
+    DWORD unwindCodeBytes = 0;
+    BYTE* pUnwindBlock = nullptr;
+#ifdef UNIX_AMD64_ABI
+    DWORD cfiCodeBytes = 0;
+    BYTE* pCfiBlock = nullptr;
```
Comment (pgavlin) [inline src/jit/unwindamd64.cpp]: Why split these out for CFI? Why not just reuse `unwindCodeBytes` and `pUnwindCodeBlock`? This would also let you remove the conditional at the end of the function.

File: src/jit/unwindamd64.cpp (hunk 20 above)
Comment (kyulee1) [inline src/jit/unwindamd64.cpp]: Code duplication is intentional to cleanly separate from Windows code per Bruce's feedback. As commented, once we introduce a unwind blob that we can virtually dispatch these things per platform/target/flavor.

**Comment by pgavlin** [comment]
LGTM modulo the comments above.

File: src/jit/unwindamd64.cpp (hunk 19 above)
Comment (kyulee1) [inline src/jit/unwindamd64.cpp]: Same comment about duplication.

File: src/jit/unwindamd64.cpp (hunk 18 above)
Comment (kyulee1) [inline src/jit/unwindamd64.cpp]: I'm not sure what the minimal bar. Many small functions need very small CFIs like 1 ~2. But complex functions with lots of eh could need many. @BruceForstall Can you suggest what the reasonable size for this vectory?

Comment (BruceForstall) [inline src/jit/unwindamd64.cpp]: Not sure. Assuming it is 1-to-1 CFI-to-Windows unwind codes, you could probably do a pretty simple textual analysis of the generated ngen-based asm diffs files to figure this out.

**Comment by kyulee1** [comment]
@BruceForstall @pgavlin Updated the change per feedbacks.
- Change banner
- Change pointer location and struct definition
- Use full expansion of cases for dwarf reg conversion
- Simplify code path by sharing local variables
- Initial vector size is 5 based on investigation on a few examples
- Didn't hoist the common part of code on purpose in order to split the code path.

**Comment by pgavlin** [comment]
LGTM. Thanks for doing this!

**Comment by BruceForstall** [comment]
LGTM

**Comment by kyulee1** [comment]
Failure seems unrelated.
